<a href="https://colab.research.google.com/github/contatofe/Estudos-e-Desafios/blob/main/Desafio_4_Regress%C3%A3o_com_redes_neurais_Udemy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Obs.: esse exercício é uma continuação do exercício 2 de regressão.

Resultados obtidos no segundo desafio:

**Regressão linear** - R2: 77.24% | RMSE: 6240.21 | R2 Validação Cruzada: 72.70%

**SVM** - R2: 83.13% | RMSE: 5373.08 | R2 Validação Cruzada: 81.35%

**Árvore de Decisão** - R2: 87.23% | RMSE: 4674.52 | R2 Validação Cruzada: 84.01%

**Random Forest** - R2: 87.48% | RMSE: 4627.77 | R2 Validação Cruzada: 84.79%

**XGBoost** - R2: 87.79% | RMSE: 4570.70 | R2 Validação Cruzada: 84.95%

**LightGBM** - R2: 87.85% | RMSE: 4560.51 | R2 Validação Cruzada: 84.80%

**Catboost** - R2: 87.31% | RMSE: 4659.89 | R2 Validação Cruzada: 84.60%

# Importações

In [ ]:
import warnings
warnings.filterwarnings("ignore")

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPRegressor

# Pré Processamento

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Arquivos - Dados/insurance.csv')

In [ ]:
df.drop_duplicates(inplace=True)

In [ ]:
previsores = df.drop(columns=['charges'])
alvo = df['charges']

In [ ]:
#Transformando variáveis categóricas em numéricas.

le = LabelEncoder()
previsores['sex'] = le.fit_transform(previsores['sex'])
previsores['smoker'] = le.fit_transform(previsores['smoker'])

previsores = pd.concat([previsores, pd.get_dummies(previsores['region'])], axis=1)
previsores.drop(columns=['region'], inplace=True)

#Separando bases de treino e teste

In [ ]:
x_treino, x_teste, y_treino, y_teste = train_test_split(previsores, alvo, test_size = 0.3, random_state = 42)

# Criando modelos

##Redes neurais

In [ ]:
#previsões
redes = MLPRegressor(hidden_layer_sizes=(10,10), activation='relu', verbose=False, max_iter=1000,solver='lbfgs')
redes.fit(x_treino, y_treino)
y_pred = redes.predict(x_teste)

In [ ]:
#métricas - teste
print(f'R2 = {(r2_score(y_teste, y_pred)):.2%}')
print(f'RMSE = {np.sqrt(mean_squared_error(y_teste, y_pred)):.2f}')
print(f'MAE = {mean_absolute_error(y_teste, y_pred):.2f}')

R2 = 85.38%
RMSE = 5001.93
MAE = 2928.50


In [ ]:
#validação cruzada
kfold = KFold(n_splits = 30, shuffle=True, random_state = 42)
modelo = redes
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)
print("R2 Médio: %.2f%%" % (resultado.mean() * 100.0))

R2 Médio: 74.63%


##Resultados

**Regressão linear** - R2: 77.24% | RMSE: 6240.21 | R2 Validação Cruzada: 72.70%

**SVM** - R2: 83.13% | RMSE: 5373.08 | R2 Validação Cruzada: 81.35%

**Árvore de Decisão** - R2: 87.23% | RMSE: 4674.52 | R2 Validação Cruzada: 84.01%

**Random Forest** - R2: 87.48% | RMSE: 4627.77 | R2 Validação Cruzada: 84.79%

**XGBoost** - R2: 87.79% | RMSE: 4570.70 | R2 Validação Cruzada: 84.95%

**LightGBM** - R2: 87.85% | RMSE: 4560.51 | R2 Validação Cruzada: 84.80%

**Catboost** - R2: 87.31% | RMSE: 4659.89 | R2 Validação Cruzada: 84.60%

**Redes Neurais** - R2: 85.38% | RMSE: 5001.93 | R2 Validação Cruzada: 74.63%

---

As redes neurais performaram um pouco pior que os modelos que já estavam vencendo.